readme.md
<br>
<br>
<p align="center"><img src="./icon.svg" width="150"></p>
<br>
<br>

```
> npm i sprut-protocol-generator
```

# Sprut-protocol-generator

<span style="color: #cc5555">Инструмент тестируется и будет дописываться или переписываться.</span>

<span style="color: #cc5555">Подробное поисание ,eltn после тестирования.</span>

---

Инструмент для описания протоколов двухстороннего взаимодействия на основе командных событий с последующим созданием типизированных буферов сообщений и перевода буферов обратно в данные.

Можно использовать с различным типом взаимодействия: tcp-соединение, в общении родительских и дочерних процессов, в буфферах с общим доступом.

---

<br>

## <span id="contents">Оглавление</span>

- [Использование Connection и MultiplexConnection](#connection)
- [Использование Server](#server)

<br>

## <span id="connection">Использование Connection и MultiplexConnection:</span>

- [К оглавлению](#contents)

<br>

```js
...
```

<br>

## <span id="server">Использование Server:</span>

- [К оглавлению](#contents)

<br>

```js
...
```

Описаны не все особенности. Некоторое из неописанного, возможно, будет выпилено.

change_log.md
## 0.0.1-beta.1 --> 

src/server.js
const { createServer } = require('net');
const { EventEmitter } = require('events');
const { Connection } = require('./connection');
const { MultiplexConnection } = require('./multiplexConnection');
const { MessageClass, VarTypes } = require("sprut-buffer-structure");

const ServerSymbols = {
    connections: Symbol('connections'),
    connectionClass: Symbol('connectionClass'),
    beginOfMessage: Symbol('beginOfMessage'),
    messageClasses: Symbol('messageClasses')
}

class Server extends EventEmitter {
    constructor(config = {}) {
        super();

        this[ServerSymbols.connections] = [];

        const {
            port = 80,
            host = 'localhost',
            path,
            byteLength,
            isMultiplex,
        } = config;

        this.port = port;
        this.host = host;
        this.path = path;

        this[ServerSymbols.beginOfMessage] = MessageClass(isMultiplex ?
            {
                dataTypes: [
                    VarTypes.UNumber(byteLength.command),
                    VarTypes.UNumber(byteLength.id)
                ]
            }
            : {
                dataTypes: [
                    VarTypes.UNumber(byteLength.command)
                ]
            }
        );
        this[ServerSymbols.connectionClass] = isMultiplex ? MultiplexConnection : Connection;
        this[ServerSymbols.messageClasses] = {};

        for(let command in byteLength.commands) {
            this[ServerSymbols.messageClasses][command] = MessageClass({ dataTypes: [ ...byteLength.commands[command] ] });
        }
    }

    serve(server) {

    }

    start(startListener = () => {}) {
        this._server = createServer(baseServerHandler.bind(this));

        if(this.path) {
    		this._server.listen(this.path, () => {
                console.log(`Server listen on path ${this.path}.`);
                startListener.apply(this);
            });
        } else {
            this._server.listen(this.port, this.host, () => {
                console.log(`Server listen on ${this.host}:${this.port}.`);
                startListener.apply(this);
            });
        }
    }

    stop() {
        this._server.close();
    }

    forceStop() {
        this[ServerSymbols.connections].forEach(connection => {
            connection.close();
        });

        this.stop();
    }

    get messageClasses() {
        return {...this[ServerSymbols.messageClasses]};
    }
}

function baseServerHandler(socket) {
    const connection = new this[ServerSymbols.connectionClass](undefined, socket, this);

    this[ServerSymbols.connections].push(connection);
    socket.on('error', (error) => console.log('error', error))

    connection.setMessageClasses(this[ServerSymbols.messageClasses]);
    connection.setBeginOfMessage(this[ServerSymbols.beginOfMessage]);
    this.emit('connectionStart', connection);
}

module.exports = {
    Server
}

src/process.js
const { EventEmitter } = require('events');
const { Counter } = require("./utils");

const ProcessSymbols = {
    isEnd: Symbol('isEnd')
};
const processErrorCounter = Counter();

const ProcessError = {
  unknownCommand: processErrorCounter()
};

class Process extends EventEmitter {
    constructor({ connection, id }) {
        super();

        this.connection = connection;
        this.id = id;
    }

    message(command, ...data) {
        if (!this[ProcessSymbols.isEnd]) {
            if (typeof command === 'number') {
                return this.connection.send(command, this.id, ...data);
            } else {
                throw new Error(`First argument must be typeof <number>`);
            }
        }
    }

    get isLive() {
        return !this[ProcessSymbols.isEnd];
    }

    kill() {
        if (!this[ProcessSymbols.isEnd]) {
            this[ProcessSymbols.isEnd] = true;
            this.emit('end');
        }
    }
}

module.exports = {
    Process,
    ProcessError
}

src/connection.js
const { EventEmitter } = require('events');
const { MessageClass, VarTypes } = require('sprut-buffer-structure');
const { connect } = require('net');
//const { serviceCommandHandler } = require("./serviceCommandHandler");

const ReservedCommand = {
    255: () => {}
};

const ConnectionError = {
    unknownCommand: 0
};

const ConnectionSymbols = {
    messageClasses: Symbol('MessageClasses'),
    beginOfMessage: Symbol('beginOfMessage')
};

class Connection extends EventEmitter {
    constructor(config = {}, socket, server) {
        super();

        this._socket = socket;
        this._server = server;
        this._handleUnknownCommand = server ? server.handleUnknownCommand : false;

        if (config.byteLength) {
            this[ConnectionSymbols.beginOfMessage] = MessageClass({
                dataTypes: [
                    VarTypes.UNumber(config.byteLength.command),
                    VarTypes.UNumber(config.byteLength.id)
                ]
            });
        }
        this[ConnectionSymbols.messageClasses] = {};

        if (config.commands) {
            for(let command in config.commands) {
                this[ConnectionSymbols.messageClasses][command] = MessageClass({ dataTypes: [ ...config.commands[command] ] });
            }
        }

        if(socket) {
            initSocketConnection(socket, this);
        }
    }

    get messageClasses() {
        return this[ConnectionSymbols.messageClasses];
    }

    get socket() {
        return this._socket;
    }

    get server() {
        return this._server;
    }

    setMessageClasses(messageClasses) {
        if(!this[ConnectionSymbols.messageClasses]) {
            this[ConnectionSymbols.messageClasses] = messageClasses;
        }
    }

    setBeginOfMessage(messageClass) {
        if(!this[ConnectionSymbols.beginOfMessage]) {
            this[ConnectionSymbols.beginOfMessage] = messageClass;
        }
    }

    send(command, ...data) {
        return this._socket.write(
            Buffer.concat([
                new this[ConnectionSymbols.beginOfMessage](command).toBuffer(),
                new this[ConnectionSymbols.messageClasses][command](...data).toBuffer()
            ])
        );
    }

    message(command, ...data) {
        return Buffer.concat([
            new this[ConnectionSymbols.beginOfMessage](command).toBuffer(),
            new this[ConnectionSymbols.messageClasses][command](...data).toBuffer()
        ]);
    }

    serveSocket(socket) {
        this._socket = socket;
        socket.on('connection', () => {
            console.log(`Connection is succesfully.`);
        });

        initSocketConnection(this._socket, this);
    }

    connect({
        port = 80,
        host = 'localhost',
        path,
        byteLength,
    }, connectListener = () => {}) {
        if(this.path) {
    		this._socket = connect(path, () => {
                console.log(`Connection to server on path ${path} is succesfully.`);
                connectListener.apply(this);
            });
        } else {
            this._socket = connect(port, host, () => {
                console.log(`Connection to server ${host}:${port} is succesfully.`);
                connectListener.apply(this);
            });
        }

        initSocketConnection(this._socket, this);
    }

    handleConnectionCommand(command, values) {
        ReservedCommand[command].apply(this, [values]);
    }

    close() {
        this._socket?.end();
    }
}

async function baseConnectionHandler(messageData) {
    if(!this[ConnectionSymbols.messageClasses]) {
        // todo
        throw new Error();
    }

    const [ baseMessageData, dataOffset ] = this[ConnectionSymbols.beginOfMessage].fromBuffer(messageData);
    const command = baseMessageData[0];

    if (!this[ConnectionSymbols.messageClasses][command]) {
        this.emit('commandError', `Unhandled message with command ${command}`, ConnectionError.unknownCommand);
        return;
    }

    const [ dataValues ] = this[ConnectionSymbols.messageClasses][command].fromBuffer(messageData, dataOffset);
    const eventData = {
        values: dataValues,
        rawBuffer: messageData
    };

    if (!ReservedCommand[command]) {
        this._processes[id].emit('command', command, eventData);
        this._processes[id].emit(command, eventData);
    } else {
        this.handleConnectionCommand(command, dataValues);
    }
}

function initSocketConnection(socket, connection) {
    socket.on('end', () => connection.emit('end'));
    socket.on('close', () => connection.emit('close'));
    socket.on('data', data => baseConnectionHandler.apply(connection, [data]));
    socket.on('error', (error) => console.log('error', error));
}

module.exports = {
    Connection
}

src/multiplexConnection.js
const { EventEmitter } = require('events');
const { MessageClass, VarTypes } = require('sprut-buffer-structure');
const { Process, ProcessError} = require('./process');
const { connect } = require('net');
//const { serviceCommandHandler } = require("./serviceCommandHandler");

const ReservedCommand = {
    255: () => {}
};

const ConnectionSymbols = {
    messageClasses: Symbol('MessageClasses'),
    beginOfMessage: Symbol('beginOfMessage')
};

class MultiplexConnection extends EventEmitter {
    constructor(config = {}, socket, server) {
        super();

        this._socket = socket;
        this._server = server;
        this._processes = {};
        this._handleUnknownCommand = server ? server.handleUnknownCommand : false;

        if (config.byteLength) {
            this[ConnectionSymbols.beginOfMessage] = MessageClass({
                dataTypes: [
                    VarTypes.UNumber(config.byteLength.command),
                    VarTypes.UNumber(config.byteLength.id)
                ]
            });
        }

        if (config.commands) {
            this[ConnectionSymbols.messageClasses] = {};

            for(let command in config.commands) {
                this[ConnectionSymbols.messageClasses][command] = MessageClass({ dataTypes: [ ...config.commands[command] ] });
            }
        }

        if(socket) {
            initSocketConnection(socket, this);
        }
    }

    get messageClasses() {
        return this[ConnectionSymbols.messageClasses];
    }

    get socket() {
        return this._socket;
    }

    get server() {
        return this._server;
    }

    setMessageClasses(messageClasses) {
        if(!this[ConnectionSymbols.messageClasses]) {
            this[ConnectionSymbols.messageClasses] = messageClasses;
        }
    }

    setBeginOfMessage(messageClass) {
        if(!this[ConnectionSymbols.beginOfMessage]) {
            this[ConnectionSymbols.beginOfMessage] = messageClass;
        }
    }

    send(command, id, ...data) {
        return this._socket.write(this.message(command, id, ...data));
    }

    message(command, id, ...data) {
        return Buffer.concat([
            new this[ConnectionSymbols.beginOfMessage](command, id).toBuffer(),
            new this[ConnectionSymbols.messageClasses][command](...data).toBuffer()
        ]);
    }

    serveSocket(socket) {
        this._socket = socket;
        socket.on('connection', () => {
            console.log(`Connection is succesfully.`);
        });

        initSocketConnection(this._socket, this);
    }

    tcpConnection({
        port = 80,
        host = 'localhost',
        path
    }, connectListener = () => {}) {
        if(this.path) {
    		this._socket = connect(path, () => {
                console.log(`Connection to server on path ${path} is succesfully.`);
                connectListener.apply(this);
            });
        } else {
            this._socket = connect(port, host, () => {
                console.log(`Connection to server ${host}:${port} is succesfully.`);
                connectListener.apply(this);
            });
        }

        initSocketConnection(this._socket, this);
    }

    createProcess(id) {
        initProcessConnection(new Process({ connection: this, id }), this);
    }

    handleConnectionCommand(command, values) {
        ReservedCommand[command].apply(this, [values]);
    }

    close() {
        this._socket?.end();
    }

    killProcess(id) {
        if (this._processes[id]) {
            this._processes[id].kill();
        }
    }
}

async function baseConnectionHandler(messageData) {
    if(!this[ConnectionSymbols.messageClasses]) {
        // todo
        throw new Error();
    }

    const [ baseMessageData, dataOffset ] = this[ConnectionSymbols.beginOfMessage].fromBuffer(messageData);
    const command = baseMessageData[0];
    const id = baseMessageData[1];

    if (!this._processes[id]) {
        initProcessConnection(new Process({
            connection: this,
            id,
            messageClass: this[ConnectionSymbols.messageClasses]
        }), this);
    }

    if (!this[ConnectionSymbols.messageClasses][command]) {
        this._processes[id].emit('errorRequest', `Unhandled message with command ${command}`, ProcessError.unknownCommand);
        return;
    }

    const [ dataValues ] = this[ConnectionSymbols.messageClasses][command].fromBuffer(messageData, dataOffset);
    const eventData = {
        values: dataValues,
        rawBuffer: messageData
    };

    if (!ReservedCommand[command]) {
        this._processes[id].emit('command', command, eventData);
        this._processes[id].emit(command, eventData);
    } else {
        this.handleConnectionCommand(command, dataValues);
    }
}

function initSocketConnection(socket, connection) {
    socket.on('end', () => connection.emit('end'));
    socket.on('close', () => connection.emit('close'));
    socket.on('data', data => baseConnectionHandler.apply(connection, [data]));
    socket.on('error', (error) => console.log('error', error));
}

function initProcessConnection(process, connection) {
    connection._processes[process.id] = process;
    connection.emit('processStart', process);
    process.once('end', deleteProcess.bind(undefined, connection, process));
}

function deleteProcess(connection, process) {
    delete this.killProcess(process.id);
}

module.exports = {
    MultiplexConnection
}